rust/automerge-c/src/doc.rs

```diff
@@ -386,6 +387,102 @@ pub unsafe extern "C" fn AMgetChangesAdded(doc1: *mut AMdoc, doc2: *mut AMdoc) -
     to_result(doc1.get_changes_added(doc2))
 }
 
+/// \memberof AMdoc
+/// \brief Gets an `AMcursor` i.e. a stable address for a position within a list
+///        object or text object.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \param[in] obj_id A pointer to an `AMobjId` struct or `AM_ROOT`.
+/// \param[in] position The absolute position of the cursor.
+/// \param[in] heads A pointer to an `AMitems` struct with `AM_VAL_TYPE_CHANGE_HASH`
+///                  items to select a historical object or `NULL` to select the
+///                  current object.
+/// \return A pointer to an `AMresult` struct with an `AM_VAL_TYPE_CURSOR` item.
+/// \pre \p doc `!= NULL`
+/// \pre \p position `< AMobjSize(`\p doc, \p obj_id, \p heads `)`
+/// \warning The returned `AMresult` struct pointer must be passed to
+///          `AMresultFree()` in order to avoid a memory leak.
+/// \internal
+///
+/// # Safety
+/// doc must be a valid pointer to an AMdoc
+/// obj_id must be a valid pointer to an AMobjId or std::ptr::null()
+/// heads must be a valid pointer to an AMitems or std::ptr::null()
+#[no_mangle]
+pub unsafe extern "C" fn AMgetCursor(
+    doc: *const AMdoc,
+    obj_id: *const AMobjId,
+    position: usize,
+    heads: *const AMitems,
+) -> *mut AMresult {
+    let doc = to_doc!(doc);
+    let obj_id = to_obj_id!(obj_id);
+    match heads.as_ref() {
+        None => {
+            return to_result(doc.get_cursor(obj_id, position, None));
+        }
+        Some(heads) => match <Vec<am::ChangeHash>>::try_from(heads) {
+            Ok(heads) => {
+                return to_result(doc.get_cursor(obj_id, position, Some(heads.as_slice())));
+            }
+            Err(e) => {
+                return AMresult::error(&e.to_string()).into();
+            }
+        },
+    };
+}
+
+/// \memberof AMdoc
+/// \brief Gets the absolute position of an `AMcursor` within a list object or
+///        text object.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \param[in] obj_id A pointer to an `AMobjId` struct or `AM_ROOT`.
+/// \param[in] cursor A pointer to an `AMcursor` struct.
+/// \param[in] heads A pointer to an `AMitems` struct with `AM_VAL_TYPE_CHANGE_HASH`
+///                  items to select a historical object or `NULL` to select the
+///                  current object.
+/// \return A pointer to an `AMresult` struct with an `AM_VAL_TYPE_UINT` item.
```

Comment:
Perhaps this is the wrong place for it, but it would be helpful to specify the units of the return value. Is it measuring code points, bytes, or something else?

Reply:
Since PR #641, indices are assumed to be in terms of UTF-8 characters but you can build a version that uses UTF-32 indexing using the `AUTOMERGE_C_UTF32` compile time flag.

Reply:
> To be clear, indices are byte positions within a UTF8 stream by default?

@whereswaldon, for an `AM_ OBJ_TYPE_TEXT` object, yes. For an ` AM_OBJ_TYPE_LIST` object it's always the index of a value regardless of how the library is compiled.

Reply:
Makes sense, thanks for clarifying.